File: algorithms/medium/construct_smallest_number_from_di_string.ex
defmodule Solution do
  @spec smallest_number(pattern :: String.t()) :: String.t()
  def smallest_number(pattern) do
    pattern
    |> String.graphemes()
    |> aux(?1, [], [])
    |> List.to_string()
  end

  def aux([], n, st, ans) do
    ans ++ [n | st]
  end

  def aux([c | cs], n, st, ans) do
    st = [n | st]

    case c do
      "I" -> aux(cs, n + 1, [], ans ++ st)
      "D" -> aux(cs, n + 1, st, ans)
    end
  end
end

File: algorithms/medium/construct_smallest_number_from_di_string.cpp
class Solution {
public:
    string smallestNumber(string pattern) {
      string ans;
      int j = 0;
      for (int i = 0; i <= pattern.size(); ++i) {
        ans.push_back('1' + i);
        if (i == pattern.size() || pattern[i] == 'I') {
          reverse(ans.begin() + j, ans.end());
          j = i + 1;
        }
      }
      return ans;
    }
};
